Increase session gc lifetime

markstory opened the pull request:
When sessions have serverside timeouts disabled we should set a longer session.gc_maxlifetime option as using 0 could result in all sessions being reaped when a GC sweep is done.

Refs #17513

markstory committed:
Increase session gc lifetime

When sessions have serverside timeouts disabled we should set a longer
session.gc_maxlifetime option as using 0 could result in all sessions
being reaped when a GC sweep is done.

Refs #17513

othercorey commented on src/Http/Session.php:
```diff
@@ -220,8 +220,14 @@ public function __construct(array $config = [])
+            if ($maxlifetime === 0) {
+                // If sessions are set to have no idle timeout, extend the
+                // gc_maxlifetime to 30 days so that sessions don't get reaped immediately
+                $maxlifetime = 60 * 60 * 24 * 30;
```
Wonder if we should print a notice here instead of only hidden defaults so users are force to configure.

We'll have to be very clear in the docs or users will assume we're forcing this lifetime.

markstory replied:
I'm not sure this is the right solution either, but setting gc_maxlifetime to 0 is certainly wrong. We could check the current ini option value and use that as a default instead :shrug: That would at least remove the hidden fallback value.

othercorey replied:
A notice would requires users to configure it through cake instead of directly in their php ini config which does feel incorrect.

markstory replied:
I pushed up another commit that I think will be safer. When `Session.timeout` is set > 0 the behavior is the same as now, but when `Session.timeout` is `0` then Cake won't change PHP configuration and users would get the maxlifetime defined in their PHP configuration.

othercorey replied:
Looks ok. Basically, if you set a time out,  the gc will happen at the same time.

markstory committed:
refine timeout and gclifetime code

Instead of special casing 0 with a hardcoded default, we can leave the
gc_maxlifetime alone and only define it when a timeout has been set.

othercorey reviewed (approved):
Just needs the typecast.

utdrmac commented:
Shouldn't `$lifetime = 0` have a better default? If Session.timeout is not configured, then $_lifetime will be 0 and _timedOut() will always return false.

markstory commented:
> Shouldn't $lifetime = 0 have a better default? If Session.timeout is not configured, then $_lifetime will be 0 and _timedOut() will always return false.

I'm reluctant to change the defaults on something as sensitive as session management. Both `null` and `0` result in sessions with no server enforced idle timeout, but `null` doesn't change other session configuration settings.

markstory committed:
Add cast
